### PR TITLE
Add support for Microshift in CI

### PR DESCRIPTION
**Pull request by vprashar2929**
Since we have added support for having microshift as a cluster in the [local-dev-cluster](https://github.com/sustainable-computing-io/local-dev-cluster/pull/18) repo and updated the [kepler-action](https://github.com/sustainable-computing-io/kepler-action/pull/48) as well This PR is in continuation to add support for running Kepler integration tests on microshift cluster.

**Dependency:**
There is an open [PR](https://github.com/sustainable-computing-io/kepler-action/pull/51) in Kepler-action to update some recent [changes](https://github.com/sustainable-computing-io/local-dev-cluster/pull/20) done to local-dev-cluster repo.

CC: @SamYuan1990 @rootfs

### CODE REVIEW

**Comment by rootfs** [comment]
@vprashar2929 new microshift integration test failed:

https://github.com/sustainable-computing-io/kepler/actions/runs/5437485041/jobs/9888072090?pr=759
```
+ for opt in ${DEPLOY_OPTIONS}
+ unset CI_DEPLOY
+ for opt in ${DEPLOY_OPTIONS}
+ unset OPENSHIFT_DEPLOY
+ echo 'Done ./hack/build-manifest.sh'
make[1]: Leaving directory '/home/runner/work/kepler/kepler'
Error response from daemon: No such container: microshift
make: *** [Makefile:263: cluster-sync] Error 1
Error: Process completed with exit code 2.
```

**Comment by vprashar2929** [comment]
Yes, currently it will fail as there is an open [PR](https://github.com/sustainable-computing-io/kepler-action/pull/51)  in Kepler-action for updating the logic. Once that is merged then CI will work fine

File: .github/workflows/integration_test.yml [hunk 1]
```diff
@@ -8,41 +8,31 @@ jobs:
...
     steps:
       - uses: actions/checkout@v3
       - uses: actions/setup-go@v3
         with:
           go-version: 1.18
 
-      - name: use kepler action for kind cluster build
-        uses: sustainable-computing-io/kepler-action@v0.0.0
+      - name: use kepler action for cluster build
+        uses: sustainable-computing-io/kepler-action@main
```
Comment (SamYuan1990) [inline .github/workflows/integration_test.yml:19]: @rootfs, @vprashar2929  for dev branch, we should keep this to main, for release we should use a release branch. Hence before we bump up version for kepler-action and local-dev-cluster, we should keep it as v0.0.0 and keep this PR open for review.

File: .github/workflows/integration_test.yml [hunk 2]
```diff
@@ -8,41 +8,31 @@ jobs:
...
-            IMAGE_TAG: "devel"
-            CTR_CMD: docker
-            CI_ONLY: "true"
+          CLUSTER_PROVIDER: ${{matrix.cluster_provider}}
+          IMAGE_REPO: "localhost:5001"
+          IMAGE_TAG: "devel"
+          CTR_CMD: docker
+          CI_ONLY: "true"
 
-      - name: test if kepler is still alive
```
Comment (SamYuan1990) [inline .github/workflows/integration_test.yml:32]: I hope we can keep this part of check. As for historical reason,  kepler is not started correctly, hence this part is to check if kepler pod started.

File: hack/cluster-deploy.sh [hunk 3]
```diff
@@ -25,38 +25,27 @@ CLUSTER_PROVIDER=${CLUSTER_PROVIDER:-kubernetes}
...
+    then
+        $CTR_CMD image prune -a -f || true && $CTR_CMD restart microshift
+        wait_microshift_up
+    fi
+
     [ ! -d "${MANIFESTS_OUT_DIR}" ] && echo "Directory ${MANIFESTS_OUT_DIR} DOES NOT exists. Run make generate first."
     
+    if [ "$CLUSTER_PROVIDER" == "microshift" ]
+    then
+        kubectl label node --all sustainable-computing.io/kepler=''
```
Comment (SamYuan1990) [inline hack/cluster-deploy.sh:38]: we should considering change https://github.com/sustainable-computing-io/kepler/blob/main/manifests/config/exporter/patch/patch-ci.yaml

File: .github/workflows/integration_test.yml (hunk 1 above)
Comment (SamYuan1990) [reply .github/workflows/integration_test.yml:19]: hence I will suggestion as bump up version in another PR.

File: hack/cluster-up.sh [hunk 4]
```diff
@@ -20,7 +20,7 @@ if [ -d local-dev-cluster ]
 then
 	echo "use local local-dev-cluster"
 else
-    echo "download local-dev-cluster with version v0.0.0"
-	git clone -b v0.0.0 https://github.com/sustainable-computing-io/local-dev-cluster.git --depth=1
+    echo "download local-dev-cluster with latest version"
+	git clone https://github.com/sustainable-computing-io/local-dev-cluster.git --depth=1
```
Comment (SamYuan1990) [inline hack/cluster-up.sh:24]: related with version bump up. Same comments here for version control.

**Review by SamYuan1990 — CHANGES_REQUESTED** [review]
I will suggestion as bump up version in another PR.
As split this PR into 2 parts.
1. as adjust for logic, as add new verify.sh shell
2. bump up version to add microshift support.

**Comment by SamYuan1990** [comment]
to go with a new release or not for https://github.com/sustainable-computing-io/local-dev-cluster as k3s support
to go with a new release or not for https://github.com/sustainable-computing-io/kepler-action/issues/50 as Prow support
@rootfs , @vprashar2929 , @jichenjc , @marceloamaral 
I suppose we should have a vote as an agreement before we merge this PR.
and version bump up should also apply for helm and operator.
cc: @jiangphcn , @husky-parul

File: .github/workflows/integration_test.yml (hunk 1 above)
Comment (vprashar2929) [reply .github/workflows/integration_test.yml:19]: Sure bumping up the version in another PR sounds good. Till that time I will keep it as v0.0.0

File: .github/workflows/integration_test.yml (hunk 2 above)
Comment (vprashar2929) [reply .github/workflows/integration_test.yml:32]: Yes, @SamYuan1990 I have added another [script](https://github.com/sustainable-computing-io/kepler/blob/92722eb34559bd48f75edc19ac32018fc1fafa55/hack/verify.sh) inside the hack folder which will test if kepler is still alive as well as launch e2e tests. It is called from [here](https://github.com/sustainable-computing-io/kepler/blob/92722eb34559bd48f75edc19ac32018fc1fafa55/hack/cluster-deploy.sh#L48)

Comment (SamYuan1990) [reply .github/workflows/integration_test.yml:32]: ok

**Comment by vprashar2929** [comment]
One more thing I would like to add is when I was working on adding micro shift to CI I realized that building Kepler image takes more than 10GB of disk space(Dependent Images) on GH runner and we are using the same runner to create cluster, build image and deploy Kepler. Since the default [disk size](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) for GH runner is around 14GB. This make's difficult for the running cluster to run properly as I have seen the case where the cluster had to evict the running pods in order to reclaim the disk space because of `NodeHasDiskPressure`. To resolve this issue as of now I have added a [step](https://github.com/sustainable-computing-io/kepler/blob/92722eb34559bd48f75edc19ac32018fc1fafa55/hack/cluster-sync.sh#L43) to stop the cluster build Kepler image push it to the local registry and then [clean the images](https://github.com/sustainable-computing-io/kepler/blob/92722eb34559bd48f75edc19ac32018fc1fafa55/hack/cluster-deploy.sh#L30) from the runner so that there is some disk space available and then restart the cluster. I am not a fan of this approach but I was thinking can we build the Kepler image on a separate runner upload it as an artifact and then download it on a different runner where we will launch our cluster and push it to the local registry. After that follow the normal process to deploy Kepler and trigger e2e tests?

@rootfs @SamYuan1990 @marceloamaral Any thoughts on this?

**Comment by rootfs** [comment]
@vprashar2929 that sounds a good idea. Would you like to have multiple jobs, kind/microshift/and potentially k3s in different jobs? How about we divide the flow into image build and image test? The image build job builds and uploads kepler image artifacts and the image test steps download the image and run kind/microshift/k3s in their runners.

**Comment by vprashar2929** [comment]
Yes, @rootfs  I am also thinking of the same flow to have a job that builds and uploads Kepler image artifacts and another which uses our Kepler-action to bring up kind/microshift/k3s clusters, deploy Kepler, and run e2e test's on it.

**Review by rootfs — APPROVED** [review]

**Comment by rootfs** [comment]
@SamYuan1990 is the request done?

**Comment by vprashar2929** [comment]
The current integration test will fail as I have reverted the bump up's in version for sustainable-computing-io/kepler-action. The v0.0.0 doesn't contain the functionality to launch microshift cluster.

**Comment by rootfs** [comment]
@vprashar2929 got a CI failure

image

**Pull request merged**